examples/2_intermediate/mta.ipynb

```diff
@@ -0,0 +1,495 @@
+      "2024-04-17 13:06:08.873762 MTA data:Event( train={\"053650_3..N01R\", IN_TRANSIT_TO 250N @10:05:56}, update={ID: 128N, Arr: 10:42:08, Dep: 10:42:08, Sched: T3, } )\n",
+      "2024-04-17 13:06:08.873812 MTA data:Event( train={\"054100_3..N01R\", IN_TRANSIT_TO 253N @10:05:46}, update={ID: 128N, Arr: 10:48:46, Dep: 10:48:46, Sched: T3, } )\n",
+      "FetchTrainDataAdapter::stop\n",
+      "Done.\n"
+     ]
+    }
+   ],
+   "source": [
+    "import csp\n",
+    "from csp.impl.pushadapter import PushInputAdapter\n",
+    "from csp.impl.wiring import py_push_adapter_def\n",
+    "\n",
+    "import nyct_gtfs\n",
+    "\n",
+    "import os\n",
+    "import time\n",
+    "import threading\n",
+    "from datetime import datetime, timedelta\n",
+    "\n",
+    "\n",
+    "class Event(csp.Struct):\n",
+    "    train: nyct_gtfs.trip.Trip\n",
+    "    update: nyct_gtfs.stop_time_update.StopTimeUpdate\n",
+    "\n",
+    "\n",
+    "class FetchTrainDataAdapter(PushInputAdapter):\n",
+    "    def __init__(self, interval):\n",
+    "        self._interval = interval\n",
+    "        self._thread = None\n",
+    "        self._running = False\n",
+    "\n",
+    "    def start(self, starttime, endtime):\n",
+    "        print(\"FetchTrainDataAdapter::start\")\n",
+    "        self._running = True\n",
+    "        self._thread = threading.Thread(target=self._run)\n",
+    "        self._thread.start()\n",
+    "\n",
+    "    def stop(self):\n",
+    "        print(\"FetchTrainDataAdapter::stop\")\n",
+    "        if self._running:\n",
+    "            self._running = False\n",
+    "            self._thread.join()\n",
+    "\n",
+    "    def _run(self):\n",
+    "        # This is where we will read and process the real-time data feed\n",
+    "        feed = nyct_gtfs.NYCTFeed(\"https://api-endpoint.mta.info/Dataservice/mtagtfsfeeds/nyct%2Fgtfs\", api_key=\"\")\n",
+    "\n",
+    "        while self._running:\n",
+    "            print(f\"{datetime.utcnow()}: refreshing MTA feed\")\n",
+    "            feed.refresh()\n",
+    "            # trains will contain all trains underway currently headed to 34St-Penn Station.\n",
+    "            trains = feed.filter_trips(underway=True, headed_for_stop_id=['128', '128N'])\n",
+    "            # tick whenever feed is refreshed\n",
+    "            for train in trains:\n",
+    "                for update in train.stop_time_updates:\n",
+    "                    if update.stop_id in ['128', '128N']:\n",
+    "                        self.push_tick(Event(train=train, update=update))\n",
+    "            time.sleep(self._interval.total_seconds())\n",
+    "\n",
+    "FetchTrainData = py_push_adapter_def(\"FetchTrainData\", FetchTrainDataAdapter, csp.ts[Event], interval=timedelta)\n",
+    "\n",
+    "@csp.graph\n",
+    "def mta_graph():\n",
+    "    print(\"Start of graph building\")\n",
+    "    trains_at_penn = FetchTrainData(timedelta(seconds=30))\n",
```

Comment:
I would add a node after `FetchTrainData` just to pretty-print the info (like you have as a table above).
Also I'd add a delimiter line, like `----...` at the end of each table to make it clearer.